src/com/gfredericks/schpec/number.clj

```diff
@@ -0,0 +1,119 @@
+
+(defn real?
+  [x]
+  (and (number? x)
+       (case x
+         Double/POSITIVE_INFINITY false
+         Double/NEGATIVE_INFINITY false
+         Double/NaN false
+         true)))
+
+(s/def ::real
+  (s/spec real?
+          :gen #(gen/double* {:infinite? false :NaN? false})))
+
+(defn decimal-pred
+  [precision scale]
+  (fn [d]
+    (let [d (bigdec d)]
+      (and (or (not precision)
+               (>= precision (.precision d)))
+           (or (not scale)
+               (let [d-scale (.scale d)]
+                 (and (not (neg? d-scale))
+                      (>= scale d-scale))))))))
+
+(defn decimal-in
+  "Specs a decimal number. The number type may be anything that bigdec
+   accepts. Options:
+
+    :precision - the number of digits in the unscaled value (default none)
+    :scale     - the number of digits to the right of the decimal (default none)
+    :min       - minimum value (inclusive, default none)
+    :max       - maximum value (inclusive, default none)
+
+   A decimal satifies this spec if its precision and scale are not greater
+   than the specified precision and scale, if given.
+
+   Note that the java math definition of precision and scale may not be the
+   same as e.g. your database. For example, -1E-75M has a precision of 1 and a
+   scale of 75. For sanest results, you should specify both, though the spec
+   does not require both."
+  [& options]
+  (let [{:keys [precision scale min max]} options
+        dec-pred (decimal-pred precision scale)]
+    (letfn [(pred [d]
+              (and (dec-pred d)
+                   (or (not min)
+                       (>= d min))
+                   (or (not max)
+                       (>= max d))))
+            (gen []
+              (let [min (or min
+                            (and precision
+                                 (-> BigDecimal/ONE
+                                     (.movePointRight precision)
+                                     dec
+                                     .negate)))
+                    max (or max
+                            (and precision
+                                 (-> BigDecimal/ONE
+                                     (.movePointRight precision)
+                                     dec)))
+                    mc (when precision
+                         (MathContext. precision RoundingMode/HALF_UP))]
+                (letfn [(f [d]
+                          (cond-> (bigdec d)
+                            scale
+                            (.setScale scale BigDecimal/ROUND_HALF_UP)
+                            precision
+                            (.round mc)))]
+                  (gen/fmap f (gen/double* {:infinite? false
+                                            :NaN? false
+                                            :min min
+                                            :max max})))))]
+      (s/spec pred :gen gen))))
+
+(s/def :specs.number.decimal/precision
```

Comment:
Yeah, they're not intended as published specs for general use, they're just to support the s/keys :opt-un. I've tried a few different techniques for bespoke options specs and haven't really found anything I like better than this yet.